### ISSUE TITLE
Andrii : Behaviour, Strategy, Implementation, 3 weeks 

### ISSUE THREAD

**Issue by andriivam** [lab-brussels-1/home#355]
## Learning Objectives

<details>
<summary>Priorities: 🥚, 🐣, 🐥, 🐔 (click to learn more)</summary>
<br>

There is a lot to learn in this repository. If you can't master all the material
at once, that's expected! Anything you don't master now will always be waiting
for you to review when you need it. These 4 emoji's will help you prioritize
your study time and to measure your progress:

- 🥚: Understanding this material is required, it covers the base skills you'll
  need to move on. You do not need to finish all of them but should feel
  comfortable that you could with enough time.
- 🐣: You have started all of these exercises and feel you could complete them
  all if you just had more time. It may not be easy for you but with effort you
  can make it through.
- 🐥: You have studied the examples and started some exercises if you had time.
  You should have a big-picture understanding of these concepts/skills, but may
  not be confident completing the exercises.
- 🐔: These concepts or skills are not necessary but are related to this module.
  If you are finished with 🥚, 🐣 and 🐥 you can use the 🐔 exercises to push
  yourself without getting distracted from the module's main objectives.

---

</details>

### [1. Remix](./1-remix/)

Practice studying and remixing other people's solutions to coding challenges. Create your own solutions by mixing and matching pieces from other people's code.

- 🐣 **Reconstructing**: You can reconstruct a variety of solutions to the same challenge when they are presented as parsons problems.
- 🐣 **Analyzing**: You can analyze a function written at your level. This includes:
  - [ ] **Behavior**: Write documentation, test cases and use cases to describe the function's behavior.
  - [ ] **Strategy**: Describe the function's strategy using plain english.
  - [ ] **Implementation**: List the language features in a function and explain how each one is used.
  - [ ] **Small Changes**: You can think of 2+ changes to the function's _implementation_ that would not change it's _strategy_.
- 🐣 **Remixing** You can analyze several solutions to the same challenge then ...
  - [ ] **Write**: Your own solution by remixing the ones you studied.
  - [ ] **Explain**: How the other solutions inspired yours; what ideas did you take from them? what ideas did you _not_ take?
  - [ ] **Analyze**: Complete a write-up your own solution as though someone else wrote it.

### [2. Write](./2-write/)

- 🐣 **Function Design**:
  - [ ] **Writing Tests**: Given a working function, you can write passing test cases to describe it's behavior.
  - [ ] **Writing Functions**: You can design multiple solutions to the same code challenge, keeping notes about different experiments you tried along the way.
- [ ] 🐣 **Generating Documentation**: You can write a JSDoc comment for your solutions and run a script to generate markdown documentation.
- [ ] 🐣 **Fuzz Testing**: You write solutions that pass randomly generated test cases.
- 🐣 **Test Driven Development**: You can solve open-ended, ambiguous coding challenges at your level:
  - [ ] **Reading Docs**: You can understand what the function is supposed to do by reading it's JSDoc description.
  - [ ] **Writing Tests**: You can write test cases _before_ there is a function to test.
  - [ ] **Writing Functions**: You can write one function that passes the test cases you have prepared (even if it's just 1 test case!).
  - [ ] **Refactoring**: You can improve your function's implementation _without_ failing any test cases that were passing.
  - [ ] **Iterative Development**: You can repeat the TDD process until you are satisfied with your test cases and solution.
- [ ] 🐔 **Code Golf**: Write your solutions with the fewest characters possible! This won't help you write readable code, but it will make you think deeply about JS, your strategy and implementation.

### [3. Review](./3-review/)

- 🐣 **Continuous Integration**: You can check your code's quality _before_ pushing so your CI checks all pass.
  - [ ] **Formatting**
  - [ ] **Linting**
  - [ ] **Testing**
- [ ] 🐣 **Code Review**: You can use a checklist to give a thorough, positive and constructive review of your classmates' solutions.
- [ ] 🐔 **Code Coverage**: You can explain what code coverage is, why it's important, and can write unit tests with 100% code coverage.

---

**Comment by andriivam** [comment]
# Week 1

## I Need Help With:
So far i'm don't need help with testing. Could be great if someone recommend tutorial like 'edabit'  but where also we can write not only function but test too)

## What went well?

I learned the syntax for testing. At the beginning its looks scary and difficult but after i wrote couple of test i understand well) 

## What went less well?

For the moment hard to find best case for test. And always think did i choose the best cases, do i need add something or change. Also i had some issue when i need to negative test like throw error (i made mistake with syntax, didn't wrap code into function).

## Lessons Learned

Writing and passing test helps not check is function work correct but its really help to understand how is function work.

## Sunday Prep Work
Study the repo and trying to write some test for other functions

**Comment by danielhalasz** [comment]
> Could be great if someone recommend tutorial like 'edabit' but where also we can write not only function but test too)

let's ask Thibault, I do not know of any such sites.

> I learned the syntax for testing. At the beginning its looks scary and difficult but after i wrote couple of test i understand well)

💪 



> For the moment hard to find best case for test.

yes, you are right, writing good tests is not easy! it really makes you think about all the possible situations!

**Comment by danielhalasz** [comment]
hopefully it is more clear after our Slack discussion

**Comment by andriivam** [comment]
@danielhalasz Absolutely. It helps me a lot. really appreciate  for all your help and explanation.

**Comment by andriivam** [comment]
# Week 3
[Group 6 practice code review ](https://github.com/alihaidermalik20/practice-code-review)

## I Need Help With:
-So far i don't need help but some task looks really complex and sometimes not easy to find example in internet to get clue what i need to do.
## What went well?
- Writing the solution for our group project and formatting the code. I hope with PR i also will not have problem)
## What went less well?
- As i didn't try yet to push my solution but im pretty sure there will be some problems based on my experience when. we were doing exercises in the class. Sometimes it can be really tricky  and even not logical (in the class i had problem as `npm run format` fights with `prettier`) so i couldn't pass the checks during when i was doing my PR as i always have lint issues.
## Lessons Learned
- Sometimes it couldn't be not really easy as it look from begging to format well you repo and push it but its really important as well formatted  code much more easy to read plus in real work we always need to follow some rules which provides company. Also now (of course using internet) im able to write solution for much more difficult exercises.
## Sunday Prep Work
- We divided task by persons but as there no class in Sunday i do all challenges in my own repo

**Comment by andriivam** [comment]
@danielhalasz As soon i will get access to my group repo i will do it my PR. for the moment all looks pretty cool so i hope will not have any issues during my PR)

**Comment by andriivam** [comment]
@danielhalasz  As well would like to say thank you for all your help and support)And answering all my question))i have a lot of them)
